### PR TITLE
Releases > Include PRs with complex testing instructions in a separate section within the Release Notes

### PR DESCRIPTION
**Pull request by nefeline**
This is a proposal for us to create a new section in the Release Testing instructions called "Testing notes for the development team" and include all PRs originally flagged as `Do not include in the Testing Notes`. This way, the Porters can test these PRs before the release instead of just skipping them altogether or relying exclusively on simpler regression testing for such cases.

This PR also adds an extra line in the release and patch instructions to make it clear to the Release Manager that if a PR has the `Experimental` flag, it should not be included in the testing.

### The Motivation

It is not uncommon for us to open PRs that require code changes for testing; a recent example is: https://github.com/woocommerce/woocommerce-blocks/pull/10128

Traditionally, these types of PRs have been marked as `Do not include in the Testing Notes` with the primary motivation being that end users cannot test them, but the changes made on such cases can impact end users as much as any other PRs with less complex test instructions. In other words, the idea here is that: complex testing instructions are not a valid reason for a PR not to be tested before a release.

With that said, to bring more consistency and increase our confidence in the release process even further, the changes proposed on this PR are:

- In the Pull Request template, the `Do not include in the Testing Notes` check was renamed to `Should be tested by the development team exclusively`.
- When creating the Testing Notes for the release, the Release Manager now adds the instructions for such PRs in a new separate section called "Testing notes for the development team".
- The Porters should then proceed to test these changes as any other PRs in the release.

Slack convo about this topic for reference: p1689738581641489/1689603956.913189-slack-C02UBB1EPEF

Tagging @tjcafferkey , @ralucaStan, and @danielwrobert for your thoughts on this as well. Thanks in advance!

### CODE REVIEW

**Comment by kmanijak** [comment]
Thanks for raising this topic and this PR! 👍 

I agree with the suggested changes, they make sense to me!

Although, one problem I see is the testing itself. We test releases on a ZIP. If the testing requires changes in code, does it mean porters have to generate different ZIP or do we agree that they can test on a local build using `npm run build:deploy`? Also, I think it may be more troublesome when testing in Core.

Don't get me wrong - I still think it's a good idea to include those tests, but we need to acknowledge that actual testing will happen on a different build than the actual release ZIP. Which I still believe brings more value than omitting those tests.

**Review by Aljullu — COMMENTED** [review]
Thanks for leading this, @nefeline. A +1 from my side.

> - In the Pull Request template, the `Do not include in the Testing Notes` check was renamed to `Should be tested by the development team exclusively`.

I think there is still the case that we might want to exclude some PRs from the testing notes, ie: PRs only making changes to tests. How do you think we should mark those PRs with this new flow?

**Comment by tjcafferkey** [comment]
I think the motivation behind this change is valid and they should be tested at the time of release like any other PR and not dismissed because the PR's testing instructions are considered _complex_.

That said, there is a wider conversation to be had about duplicating effort around testing (at the point of merge, and at the point of release) I believe, which is why we started to gather feedback around our process.

I think in the context of our current process this is a welcomed improvement. Thanks @nefeline !

**Comment by ralucaStan** [comment]
> Traditionally, these types of PRs have been marked as Do not include in the Testing Notes with the primary motivation being that end users cannot test them, 

The testing notes are actually re-used when the WC Blocks version is increased in WooCommerce when the build goes through testing. 

I see renaming `Do not include in the Testing Notes` as unnecessary. 

The issue here is that these excluded PRs are not tested from what I understand. I like the idea to create an explicit section. `Should be tested by the development team exclusively`. We can instruct only the RM to include here the PRs that have `Do not include in the Testing Notes`

+1 on the experimental addition.

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -45,7 +45,7 @@ Fixes #
 2.
 3.
 
-* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->
+* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->
```
Comment (opr) [inline .github/pull_request_template.md:50]: Should we keep both boxes? @Aljullu made a good point about PRs not affecting the release, e.g. tests or build-related PRs

**Review by opr — CHANGES_REQUESTED** [review]
The changes look good, thanks for taking the initiative here! I added a comment that I think we should implement before shipping this change.

**Comment by nefeline** [comment]
Thanks all for the input! Replying to the comments:

@kmanijak regarding:

> but we need to acknowledge that actual testing will happen on a different build than the actual release ZIP

Not necessarily: we can still upload the ZIP file to any JN site, and for PRs that require a code snippet for testing, we can either edit the files directly via `Tools > Plugin File Editor` and/or `Tools > Theme File Editor` or install a plugin such as Code Snippets to insert the custom code. This way, we can guarantee that the zip file is still used, independent of what is being tested.

@Aljullu regarding:

> I think there is still the case that we might want to exclude some PRs from the testing notes, ie: PRs only making changes to tests. How do you think we should mark those PRs with this new flow?

That's true, good point! I think it makes sense to preserve the pre-existing `Do not include in the Testing Notes` check for such cases: pushing a commit to re-introduce it 👍 

@tjcafferkey regarding:

> I think in the context of our current process this is a welcomed improvement. 

👍 thanks!

@ralucaStan regarding:

> The testing notes are actually re-used when the WC Blocks version is increased in WooCommerce when the build goes through testing.

Interesting! Just to make sure I understand this correctly: does this mean that even when a PR is marked as `Do not include in the Testing Notes`, it is still tested prior to bumping the version of WC Blocks in the core of WooCommerce?

> I see renaming Do not include in the Testing Notes as unnecessary.
> The issue here is that these excluded PRs are not tested from what I understand. I like the idea to create an explicit section. Should be tested by the development team exclusively. We can instruct only the RM to include here the PRs that have Do not include in the Testing Notes

Instead of renaming, I'm updating this PR to preserve it alongside the `Should be tested by the development team exclusively` check, as Albert pointed out, there could be cases where we are just making changes to tests, for example.

> +1 on the experimental addition.

👍

@opr regarding:

> I added a comment that I think we should implement before shipping this change.

Sounds good, thanks for the review! I'm doing the changes as suggested 👍

File: .github/pull_request_template.md (hunk 1 above)
Comment (nefeline) [reply .github/pull_request_template.md:50]: Makes sense, updated 👍 !

**Comment by ralucaStan** [comment]
> Interesting! Just to make sure I understand this correctly: does this mean that even when a PR is marked as Do not include in the Testing Notes, it is still tested prior to bumping the version of WC Blocks in the core of WooCommerce?

No, it's not tested from what I know; but people are looking for this `Do not include in the Testing Notes` label

> Instead of renaming, I'm updating this PR to preserve it alongside the Should be tested by the development team exclusively check, as Albert pointed out, there could be cases where we are just making changes to tests, for example.

I think we need to make sure that not all PRs under ` Should be tested by the development team exclusively ` are tested at the release time; so the skip changelog label should still be used for PRs with tests, automation, release changes or repo hygiene changes. Issues with these changes surface usually immediately after merge, long before a release. So let's keep an eye of this in the wild to see how the teams receive it.

**Review by ralucaStan — APPROVED** [review]
Approving! Love the initiative Patricia!

**Comment by nefeline** [comment]
Thanks @ralucaStan!

> No, it's not tested from what I know; but people are looking for this Do not include in the Testing Notes label

Gotcha! Makes sense 👍 . This check was preserved as requested.

> I think we need to make sure that not all PRs under Should be tested by the development team exclusively are tested at the release time; so the skip changelog label should still be used for PRs with tests, automation, release changes or repo hygiene changes. Issues with these changes surface usually immediately after merge, long before a release. So let's keep an eye of this in the wild to see how the teams receive it.

💯  Agreed

**Review by danielwrobert — APPROVED** [review]
> and for PRs that require a code snippet for testing, we can either edit the files directly via Tools > Plugin File Editor and/or Tools > Theme File Editor or install a plugin such as Code Snippets to insert the custom code.

@nefeline JN sites also provide SSH and SFTP access so we can make code adjustments directly that way as well!

Thanks for taking the initiative to work on this - approved! 🚀

**Comment by nefeline** [comment]
> JN sites also provide SSH and SFTP access so we can make code adjustments directly that way as well!

Good point! 

> Thanks for taking the initiative to work on this - approved! 🚀

🚀

**Pull request merged**